src/engine/Eval_generic.ml

```diff
@@ -229,9 +230,96 @@ let eval_regexp_matches ?(base_offset = 0) ~file ~regexp:re str =
+           in
+           let pos = Pos.make ~file:!!file ~line ~column bytepos in
+           let loc2 = { Tok.str; pos } in
+
+           (* the names of all capture groups within the regexp *)
+           let names = Pcre.names regexp.regexp |> Array.to_list in
+           (* return regexp bound group $1 $2 etc *)
+           let n = Pcre.num_of_subs sub in
+           (* TODO: remove when we kill numeric capture groups *)
+           let numbers_env =
+             match n with
+             | 1 -> []
+             | _ when n <= 0 -> raise Impossible
+             | n ->
+                 List_.enum 1 (n - 1)
+                 |> List_.map_filter (fun n ->
+                        try
+                          let bytepos, _ = Pcre.get_substring_ofs sub n in
+                          let str = Pcre.get_substring sub n in
+                          let line, column =
+                            Xpattern_matcher.line_col_of_charpos file bytepos
+                          in
+                          let pos =
+                            Pos.make ~file:!!file ~line ~column bytepos
+                          in
+                          let loc = { Tok.str; pos } in
+                          let t = Tok.tok_of_loc loc in
+                          Some (spf "$%d" n, MV.Text (str, t, t))
+                        with
+                        | Not_found ->
+                            Logs.debug (fun m ->
+                                m ~tags "not found %d substring of %s in %s" n
+                                  regexp.pattern matched_str);
+                            None)
+           in
+           let names_env =
+             names
+             |> List_.map_filter (fun name ->
+                    try
+                      (* TODO: make exception-free versions of the missing
+                         functions in SPcre. *)
+                      let bytepos, _ =
+                        Pcre.get_named_substring_ofs regexp.regexp name sub
+                      in
+                      let bytepos = bytepos + base_offset in
+                      let str =
+                        Pcre.get_named_substring regexp.regexp name sub
+                      in
+                      let line, column =
+                        Xpattern_matcher.line_col_of_charpos file bytepos
+                      in
+                      let pos = Pos.make ~file:!!file ~line ~column bytepos in
+                      let loc = { Tok.str; pos } in
+                      let t = Tok.tok_of_loc loc in
+                      Some (spf "$%s" name, MV.Text (str, t, t))
+                    with
+                    | Not_found ->
+                        Logs.debug (fun m ->
+                            m ~tags "not found %s substring of %s in %s" name
+                              regexp.pattern matched_str);
+                        None)
+           in
+           ((loc1, loc2), names_env @ numbers_env))
```

Comment:
Is this whole block duplicated? Can we avoid it since it's not trivial? It could be done by passing a record or a first-class module containing the regexp functions as a parameter to `Xpattern_match_regexp.regexp_matcher`.

Reply:
Yes, I was a bit hesitant to add a first class module, but I do think that's better.